File: generated-code/haskell/codecraft/TcpReadWrite/src/Codegame/MessageGameModel.hs
module Codegame.MessageGameModel where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Codegame.ClientMessage (ClientMessage)
import Codegame.ServerMessage (ServerMessage)
import Data.Int

-- | Client message
data MessageGameModelClient = MessageGameModelClient {
    -- | Message
    message :: ClientMessage }
    deriving Show

instance Trans MessageGameModelClient where
    read = do
        message <- Trans.read
        return MessageGameModelClient {
            message }

    write MessageGameModelClient {
        message } = do
            Trans.write message

-- | Server message
data MessageGameModelServer = MessageGameModelServer {
    -- | Message
    message :: ServerMessage }
    deriving Show

instance Trans MessageGameModelServer where
    read = do
        message <- Trans.read
        return MessageGameModelServer {
            message }

    write MessageGameModelServer {
        message } = do
            Trans.write message

-- | Client or server message
data MessageGameModel
    -- | Client message
    = Client MessageGameModelClient
    -- | Server message
    | Server MessageGameModelServer
    deriving Show

instance Trans MessageGameModel where
    read = do
        tag :: Int32 <- Trans.read
        case tag of
            0 -> Client <$> Trans.read
            1 -> Server <$> Trans.read

    write (Client value) = do
        Trans.write (0 :: Int32)
        Trans.write value
    write (Server value) = do
        Trans.write (1 :: Int32)
        Trans.write value

File: generated-code/go/codecraft/TcpReadWrite/codegame/message_game_model.go
package codegame

import "io"
import . "trans_gen_test/stream"

// Client or server message
type MessageGameModel interface {
    // Write MessageGameModel to writer
    Write(writer io.Writer)

    // Get string representation of MessageGameModel
    String() string
}

// Read MessageGameModel from reader
func ReadMessageGameModel(reader io.Reader) MessageGameModel {
    switch ReadInt32(reader) {
    case 0:
        return ReadMessageGameModelClient(reader)
    case 1:
        return ReadMessageGameModelServer(reader)
    }
    panic("Unexpected tag value")
}

// Client message
type MessageGameModelClient struct {
    // Message
    Message ClientMessage
}

func NewMessageGameModelClient(message ClientMessage) MessageGameModelClient {
    return MessageGameModelClient {
        Message: message,
    }
}

// Read Client from reader
func ReadMessageGameModelClient(reader io.Reader) MessageGameModelClient {
    var message ClientMessage
    message = ReadClientMessage(reader)
    return MessageGameModelClient {
        Message: message,
    }
}

// Write Client to writer
func (messageGameModelClient MessageGameModelClient) Write(writer io.Writer) {
    WriteInt32(writer, 0)
    message := messageGameModelClient.Message
    message.Write(writer)
}

// Get string representation of Client
func (messageGameModelClient MessageGameModelClient) String() string {
    stringResult := "{ "
    stringResult += "Message: "
    message := messageGameModelClient.Message
    stringResult += message.String()
    stringResult += " }"
    return stringResult
}

// Server message
type MessageGameModelServer struct {
    // Message
    Message ServerMessage
}

func NewMessageGameModelServer(message ServerMessage) MessageGameModelServer {
    return MessageGameModelServer {
        Message: message,
    }
}

// Read Server from reader
func ReadMessageGameModelServer(reader io.Reader) MessageGameModelServer {
    var message ServerMessage
    message = ReadServerMessage(reader)
    return MessageGameModelServer {
        Message: message,
    }
}

// Write Server to writer
func (messageGameModelServer MessageGameModelServer) Write(writer io.Writer) {
    WriteInt32(writer, 1)
    message := messageGameModelServer.Message
    message.Write(writer)
}

// Get string representation of Server
func (messageGameModelServer MessageGameModelServer) String() string {
    stringResult := "{ "
    stringResult += "Message: "
    message := messageGameModelServer.Message
    stringResult += message.String()
    stringResult += " }"
    return stringResult
}